Update 2-service-discovery.md (#3)
koushikkothagal/reveal.js: content/microservices-interview-questions/2-service-discovery.md

## Question

> # What is ServiceDiscovery?

Reviewer: ```suggestion
# What is Service Discovery?
```

--

# What is the interviewer looking for?

> - Whether you know concept of Service Discovery.
> - Since Service discovery is one of important component in microservices arcitecture, Hence interviewer will see if you came accross
>   such concept while working on your project.
> - If you are able to give some concept on this(Service Discovery) design pattern interviewer can get along to next questions on it.

Reviewer: ```suggestion
- Understanding of the design pattern
- How the pattern applies to microservices
```

--

# Answer

> - Service Discovery is Design Pattern used in MicroServices architecture.
> - Since in Microservices architecture in order your code makes Rest API call to another service
>   - your code needs to know the network location (IP address and port) of a service instance.
>   - In a traditional application running on physical hardware, the network locations of service instances are relatively static. For example, your code can read the network       locations from a configuration file that is occasionally updated.
>   - In a modern, cloud‑based microservices application, however, this is a much more difficult problem to solve.
>   - Service instances have dynamically assigned network locations. 
>   - Moreover, the set of service instances changes dynamically because of autoscaling, failures, and upgrades. Consequently,your client code needs to use a more elaborate         service discovery mechanism.
>   - Service Discovery solves this problem.

Reviewer: ```suggestion
- Service Discovery is a design pattern used in microservices architecture.
- Helps microservices dynamically locate and call dependent services
- Necessary in microservice architecture because of dynamically assigned IPs and hosts
- Helps with autoscaling, handling failures and upgrades

```

---

## Question

> # What are types of patterns/types available in Service Discovery

Reviewer: ```suggestion
# What are types of patterns/types available in Service Discovery?
```

- Pros
   - Since the client knows about the available services instances, it can make intelligent, application‑specific load‑balancing decisions such as using hashing consistently.
- Cons
   - One significant drawback of this pattern is that it couples the client with the service registry. You must implement client‑side service discovery logic for each              programming language and framework used by your service clients. 
 - The Server‑Side Discovery Pattern 
   -  Here the client service/application does not make a request to service registry.
   -  Instead The client makes a request to a service via a load balancer.
   -  The load balancer queries the service registry and routes each request to an available service instance.
   -  The AWS Elastic Load Balancer (ELB) is an example of a server-side discovery router.
   -  Kubernetes ectd is also one example.
- Pros
   - One great benefit of this pattern is that details of discovery are abstracted away from the client.
   - This eliminates the need to implement discovery logic for each programming language and framework used by your service clients.
- Cons

>     - Unless the load balancer is provided by the deployment environment, it is yet another highly available system component that you need to set up and manage. 

Reviewer: Recommend breaking the "What are the two patterns?" and "Pros and cons" as two separate questions.

Author: Thanks for suggestion :)  will break it down

---

## Question

> # Is it Mandatory to Use Service Discovery in Microservice Architecture.

Reviewer: ```suggestion
# Is it mandatory to use service discovery in microservice architecture?
```

--

# What is the interviewer looking for?
- What type of system complexity you have worked on.
- If you have unnecessarily used Service Discovery.
- Can Microservice Architecture completes without Service Discovery.

--

# Answer

- No.
- It depends on business scenario.

> - If we have large number of microservies and these microservies can be upstream or downstream, then you can use service discovery.
> - But if microservies are small in numbers and mostly independent in business context, doen not require communication with each other then not required.

Reviewer: ```suggestion
- Ideal when there are a large number of microservices
```
